lib/logstash/outputs/tcp.rb

```diff
@@ -163,18 +181,33 @@ def register
           # Now send the payload
           client_socket.syswrite(payload) if w.any?
         rescue => e
-          @logger.warn("tcp output exception", :host => @host, :port => @port,
-                       :exception => e, :backtrace => e.backtrace)
+          log_warn "client socket failed:", e, host: @host, port: @port, socket: (client_socket ? client_socket.to_s : nil)
           client_socket.close rescue nil
           client_socket = nil
           sleep @reconnect_interval
           retry
         end
       end
     end
-  end # def register
+  end
+
+  # @overload Base#receive
+  def receive(event)
+    @codec.encode(event)
+  end
+
+  # @overload Base#close
+  def close
+    return unless @client_threads
```

Comment:
since we're defining a proper `close` here, we should also handle the closing of the `@server_socket`, otherwise a pipeline restart could lead to a "Address already in use" exception